speech/tests/system.py

```diff
@@ -245,3 +247,59 @@ def test_stream_recognize_single_utterance(self):
             for results in self._make_streaming_request(
                     file_obj, single_utterance=False):
                 self._check_results(results.alternatives)
+
+    def test_long_running_recognize(self):
+
+        def _build_large_audio_file(wav, reps):
+            ''' From https://stackoverflow.com/questions/46538867/
+                             attributeerror-wave-write-instance-has-no-
+                             attribute-exit
+            '''
+            def __enter__(self):
+                return self
+            def __exit__(self, exc_type, exc_value, traceback):
+                self.close()
+
+            wave.Wave_read.__exit__ = wave.Wave_write.__exit__ = __exit__
+            wave.Wave_read.__enter__ = wave.Wave_write.__enter__ = __enter__
```

Comment:
I agreed. Use` mock.patch`.

Reply:
I would rather avoid the patching altogether by opening the file separately, as I suggest below.

Reply:
Could I subclass it?  The patching was for the content manager to open the wave file.

Reply:
@chemelnucfin Do you see my example below?  It doesn't need patching, because it passes an already-opened file object `wave.open()`.

Reply:
The context manager doesn't exist in python 2.7, but works in python 3.6

Reply:
> The context manager doesn't exist in python 2.7, but works in python 3.6

OK.  We *still* don't need to use it to get the file closed if we wrap the `wave.open` call inside the regular `file.open` / 'io.BytesIO` call:

```python
        with io.BytesIO() as out_stream:
            out_wavfile = wave.open(out_stream, 'w')
            out_wavfile.setparams(params)
            for rep in range(15):
                out_wavfile.writeframes(frames)
        content = out_stream.getvalue()
```

Reply:
Would out_wavfile get cleaned up if an exception is thrown though?

Reply:
@chemelnucfin That example creates a wrapper around an in-memory `BytesIO` instance:  there is nothing to be cleaned up.

Reply:
ok, but is there anything wrong with my approach?  Otherwise would I have to close the wave.open somewhere?  And is that only for the out_stream or both in_stream and out_stream? thanks.

Reply:
@chemelnucfin, patching `wave.Wave_{read,write}` is intrinsically poor practice (your code does nothing to restore the original values, for instance), especially if there is a simpler alternative.  My example avoids needing to open (and therefore close / clean up) a tempfile, and is therefore both simpler to understand and less likely to break.